core/src/selection/selectionQuery.h

```diff
@@ -0,0 +1,43 @@
+#pragma once
+
+#include "glm/vec2.hpp"
+#include "tangram.h"
+#include "util/variant.h"
+#include <map>
+
+namespace Tangram {
+
+class MarkerManager;
+class FrameBuffer;
+class TileManager;
+class Labels;
+class View;
+
+enum class QueryType {
+    feature,
+    marker,
+    label,
+};
+
+using QueryCallback = variant<FeaturePickCallback, LabelPickCallback, MarkerPickCallback>;
+
+class SelectionQuery {
+
+public:
+    SelectionQuery(glm::vec2 _position, QueryCallback _queryCallback, QueryType _type);
+
+    static void clearColorCache() { s_colorCache.clear(); }
+
+    void process(const View& _view, const FrameBuffer& _framebuffer, const MarkerManager& _markerManager,
+                 const TileManager& _tileManager, const Labels& _labels) const;
+
+    QueryType type() const { return m_type; }
+
+private:
+    glm::vec2 m_position;
+    QueryCallback m_queryCallback;
+    QueryType m_type;
+
+    static std::map<size_t, uint32_t> s_colorCache;
```

Comment:
`glReadPixels` is quite known for being slow (mainly why PBO started to exist), but haven't benchmarked it in here; I just assumed that reducing calls to it would be good practice. A common pattern would be querying the system for the same position for the three kind of queries.
You're right about the static storage, but why isn't it the right data structure though? It's keyed by the hash of a 2d screen position.